Added progression variable to the ramping arrival rate

LaserPhaser opened the pull request:
Added an ability to control how the arrival rate will be added
Linear - default
instant - optional

Closes #2252

<!--


  (ﾉ◕ヮ◕)ﾉ*:・ﾟ✧
  
  Thank you for your interest in contributing to the k6 project!
  
  Before you get started, we'd kindly like to ask you to read our:
    - Contribution guidelines at https://github.com/grafana/k6/blob/master/CONTRIBUTING.md
    - Code of Conduct at https://github.com/grafana/k6/blob/master/CODE_OF_CONDUCT.md
    
  Out of respect for your time, please start a discussion regarding any bigger contributions either
  in a GitHub Issue, in the community forums or in the #contributors channel of the k6 slack before you
  get started on the implementation.
  
  If you've already done all of that, you're more than welcome to proceed with your pull request.
  Thank you again for your contribution! 🙏🏼
  
  
-->

oleiade commented on lib/executor/helpers.go:
```diff
@@ -72,6 +72,16 @@ func validateStages(stages []Stage) []error {
 			errors = append(errors, fmt.Errorf("the target for stage %d shouldn't be negative", stageNum))
 		}
+
+		switch s.Progression.String {
```
Considering `s.Progression` is defined as a `null.String` I would follow the same procedure as for the other field, and at least append an error if `s.Target.Valid` is `false.

oleiade commented on lib/executor/helpers.go:
```diff
@@ -72,6 +72,16 @@ func validateStages(stages []Stage) []error {
+
+		switch s.Progression.String {
+		case "linear",
+			"constant",
```
Did you mean "instant" instead? If not, could you clarify your reasoning? My assumption is that a `constant` interpolation would either mean a linear one (as in, "a constant rate of change"), or, no movement at all, in which case, this usage is already addressed by the existing feature-set of options (although it might feel "hacky": `stages: [{duration: '1m', target: 10}, {duration: '10m', target: 10}`, would, from my understanding, perform a "constant" rate of change).

oleiade commented on lib/executor/ramping_vus.go:
```diff
@@ -54,6 +54,7 @@ func init() {
 type Stage struct {
 	Duration types.NullDuration `json:"duration"`
 	Target   null.Int           `json:"target"` // TODO: maybe rename this to endVUs? something else?
+	Progression null.String     `json:"progression"`
```
I would prefer to name it `Interpolation` but as the naming is already being discussed in [#2256](https://github.com/grafana/k6/issues/2256), I'm fine with leaving it at that for now :+1:

oleiade commented on lib/executor/ramping_arrival_rate.go:
```diff
@@ -271,24 +271,34 @@ func (varc RampingArrivalRateConfig) cal(et *lib.ExecutionTuple, ch chan<- time.
-				ch <- time.Duration(x) + stageStart
-			}
-		} else {
+		switch progression := stage.Progression.String; progression {
```
This looks like a good refactor opportunity: to have the interpolation functions being dedicated either functions or methods called by the various switch cases. One good hint from my point of view, is that the `constant` branch is basically a duplicate of the non-ramping behavior of the `linear` one.

LaserPhaser commented:
@oleiade Thank you for reply, 
It's basically my fault that I did not close this PR. 
I've already updated the K6 documentations and added the example on how to create stepped-arrival rate. 
https://k6.io/docs/examples/instant-load-increase/

Thank you!
